Use variant<> for list box data

UseAfterFree-177 opened the pull request:
This gives us
 - Well-defined sorting order, across, and between, all types.
   - The types belong to the cells, not the columns.
   - Mixed-type columns are perfectly fine (but might look a little strange when displayed).
- The sort function no longer need to parse the formatted text.
   - So parse failures cannot happen anymore.
   - So it is now _nothrow_ exception-safe (unless we make `operator<()` throw :smiling_imp: )

One downside is that when a string is held, two copies are kept -- one in `val_` and one in `fmt_`. I did an experiment where I made strings into a cute special case, but it didn't work out (more complexity, and I still had to make the extra copy at the end).

What remains is cleaning it up (better names, better places for the new types, add doc blocks), extending it to the other tables (now, only the "Controlled Gangs" table actually uses it), and removing the old column-type scheme.

UseAfterFree-177 committed:
Sketching a CListBox that holds the actual cell value and not just the
formatted string. This way we can sort etc without having to parse
the string to recover the value (and its type).

h1262216 commented:
yeah, getting the new data into the table is going to be a bot of work. If you feel motivated, feel free to rework the
`void sGirl::OutputGirlDetailString(std::string& Data, const std::string& detailName) const` function, which is pretty horrible. 

I still dislike the specific types you have for the table, I think the variant should be over something like <int, string, bool> maybe void* (also, there is a dataid field which is currently used to associate non-text data with an entry, e.g. by keeping an index into an array of things that the screen keeps). I would like to have the table be agnostic about health/age types, and let this be the responsibility of something like this `OutputGirlDetailString` function. Otherwise, you'd need a special type for basically each kind of stat that we want to display, that would be tightly coupled to the listbox code.

having a nothrow sort is nice

UseAfterFree-177 committed:
Removing the previous column type code.

UseAfterFree-177 commented:
We don't need special types for everything we want to display -- we just need them for everything we want to give its own sorting order. If there's some type where we can construct a number or string (or whatever) to put in `val_` to get the sort order we want, then let's do that for that type! All the engine needs to know is that the variant that is the sum total of all our tabular types can be sorted with `std::list<>::sort()` and `operator<()`. To me, that's loose coupling. (Yeah, being stringly typed and endure lexically sorted numbers ("1, 10, 11, 2") is another kind of loose coupling.)

I'm still looking for a better place for the variant definition -- I just put it in cListBox for convenience. If you know a better place for it, I'll happily move it. Also, some of my names could be better.

UseAfterFree-177 committed:
Removing column types from the interface XML.

We don't need it anymore.

h1262216 commented:
OK, bu than we don't need a special type for health and age, but can instead put health=0 whenever we display DEAD and age=INT_MAX whenever we display ???, right. So the variant only needs to hold std::string and int for now (and maybe float, though I don't think this is currently used)

UseAfterFree-177 commented:
Yes. That''s roughly what I found when I actually looked inside `OutputGirlDetailString()`. So I'm removing the special `Age` and `Health` types.

UseAfterFree-177 committed:
`sGirl::OutputGirlDetail()` added to return `ItemContents`.

- `sGirl::OutputGirlDetailString()` now forwards to
  `OutputGirlDetail()` and returns the `.fmt_` part.

- A helper `sGirl::OutputGirlDetail_Job()` was added -- jobs are
  complicated enough.

- The type terms `Age` and `Health` has been removed from
  `ItemData`. It turned out that we can do without them.

TODO: Consolidate the `mk_*()` lambda helpers, when we know which ones
we need everywhere.

h1262216 commented:
OK, then we agree. Maybe I've missed it by scrolling over things, but what happens currently if two entries in the same column have two different types in their variants?

UseAfterFree-177 committed:
Bugfix. Oops!

UseAfterFree-177 commented:
In that case (different term types) `variant<>` itself  sorts by the type order. So right now that would mean `string < int < bool`.

UseAfterFree-177 committed:
Updates the dungeon to use `ItemContents`.

UseAfterFree-177 committed:
Updates the transfer screen to use `ItemContents`.

UseAfterFree-177 committed:
Consolidates `mk_text()` etc into 'cListBox.h'

UseAfterFree-177 committed:
Renames the `ItemData` and `ItemContents` types.

The new names are `CellData` and `FormattedCellData`.

UseAfterFree-177 commented:
I think I'm mostly done now here; there are just a few thing left that I want to do:

- [x] Rename `OutputGirlDetail()` (to `GetDetail()` ?) It doesn't output anything (I don't think it ever did), and the class that owns it tells you its about a girl (or customer).

- [x] Find a better place for `CellData`and friends. Or maybe 'cListBox.h' really is the best place after all.

- [x] Add a term type `Error` to `CellData`. That would let us sort errors to the end, paint them red, etc.

UseAfterFree-177 committed:
Adding an error type to `CellData`.

h1262216 commented:
Can we maybe move the `OutputGirlDetail` kind of functions (with a better name) to a separate file. I would like to have the game mechanics and game ui code to be as separate as possible (yes, that is not what the current state is). So ideally, have the game exe be made up of three components `engine` -> `game mechanics` -> `ui`. If you do that, then I think the helpers can remain in `cListBox.h`. 

I think long term, I want to move the part that returns the variant data to the game class, maybe as two separate functions, like `GetIntAttribute` and `GetBoolAttribute`, which would be used for example in managing the conditions for the xml text. Then we should unify this with the functions you've added here.

UseAfterFree-177 committed:
Renamed `OutputGirlDetail()`to `GetDetail()`.

UseAfterFree-177 commented:
I think that until we templatize the `cListBox` class, we have to have `FormattedCellData` on the engine side of the fence. And over there, `cListBox.h` is as good a place as any.

Or we could add a fourth component `common/` or `shared/` and put it there.

Anyway, I think this PR is ready to merge now. If we decide on a reorg, there'll be more things than this to move around.

h1262216 commented:
my concern wasn't about `FormattedCellData` being a part of `cListBox`, but about the fact that sGirl/sDungeonGirl now have a dependency on `cListBox`. I basically just meant to move the `GetDetails` function to a separate file as a free function. That way, UI details don't leak into the sGirl class.

UseAfterFree-177 commented:
Let's see: `FormattedCellData` is properly part of the game/engine interface, so it really belongs to both. However, in the current structure, the game can #include engine headers, while the engine cannot return the favour. So, for now, `FormattedCellData` has to live somewhere in the engine.

So I'll move it to an engine header of its own ("game_interop.h"?). Later, when we have a directory `common/` (or something) for shared stuff like this, we can move it there.

h1262216 commented:
I think having `FormattedCellData` as part of the `cListBox` is fine, because it belongs to this. Its more about the division of the game mechanics vs the game ui that is not really existing yet but that I would like to establish. Some of my previous changes were removing the `Font` references from `sGirl`, so that currently this does not depend on any interface/widgets parts anymore.

UseAfterFree-177 commented:
Anyway, I'll move `FormattedCellData` and friends to its own header (still in 'engine/'). That way `cGirl` (and others) can grab the banana without getting the whole gorilla.

h1262216 commented:
OK, I guess that is fine for now. We can revisit that once we try to pull apart the other things

UseAfterFree-177 committed:
Moves `FormattedCellData` etc to its own header file.

UseAfterFree-177 commented:
Done.

h1262216 commented:
I've fixed two small things (missing `std::` and an include of cListBox where TableCell was enough), squashed and merged.

UseAfterFree-177 commented:
Thanks!
